### PR TITLE
[Driver][SYCL] Provide shortcut options for device compilation in deb…

### PR DESCRIPTION
**Pull request by mdtoguchi**
…ug mode.

When the user passes along debug and optimization disabling options on the
command line, we should also imply equivalent options to the device
compilation.

-g -O0 (Linux) and -Zi -Od (Windows) will imply -g -cl-opt-disable for
the device compilation.  This is done for spir and fpga only at this point.

Signed-off-by: Michael D Toguchi <michael.d.toguchi@intel.com>

### CODE REVIEW

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 1]
```diff
@@ -472,8 +472,28 @@ void SYCLToolChain::TranslateTargetOpt(const llvm::opt::ArgList &Args,
...
 
+static void addImpliedArgs(const llvm::Triple &Triple,
+                           const llvm::opt::ArgList &Args,
+                           llvm::opt::ArgStringList &CmdArgs) {
+  // Current implied args are for debug information and disabling of
+  // optimizations.
+  // TODO: Add support for other architectures (gen, x86_64) as those are
+  // being defined.
+  if (Triple.getSubArch() == llvm::Triple::NoSubArch ||
+      Triple.getSubArch() == llvm::Triple::SPIRSubArch_fpga) {
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/SYCL.cpp:483]: To clarify, does this have to do with ocloc/opencl-aot's CLI requiring a separate flag for build options?

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp:483]: Yes, -g is passed to both host and device FE calls.

As for your second question, the separate flag for build options is part of the reason why I'm only doing this for device and fpga only.  It is the case, but they haven't solidified on behaviors yet so passing them now is pre-mature.  A new change will need to be done when those targets are ready.

File: clang/test/Driver/sycl-offload.c [hunk 2]
```diff
@@ -679,6 +679,13 @@
 // RUN:   | FileCheck -check-prefix=CHK-TOOLS-OPTS %s
 // CHK-TOOLS-OPTS: clang-offload-wrapper{{.*}} "-compile-opts=\"-DFOO1 -DFOO2\""
 
+/// Check for implied options (-g -O0)
+// RUN:   %clang -### -target x86_64-unknown-linux-gnu -fsycl -fsycl-targets=spir64-unknown-unknown-sycldevice -g -O0 -Xsycl-target-backend "-DFOO1 -DFOO2" %s 2>&1 \
+// RUN:   | FileCheck -check-prefix=CHK-TOOLS-IMPLIED-OPTS %s
+// RUN:   %clang_cl -### -fsycl -fsycl-targets=spir64-unknown-unknown-sycldevice -Zi -Od -Xsycl-target-backend "-DFOO1 -DFOO2" %s 2>&1 \
+// RUN:   | FileCheck -check-prefix=CHK-TOOLS-IMPLIED-OPTS %s
+// CHK-TOOLS-IMPLIED-OPTS: clang-offload-wrapper{{.*}} "-compile-opts=\"-g -cl-opt-disable -DFOO1 -DFOO2\""
+
```
Comment (AGindinson) [inline clang/test/Driver/sycl-offload.c:688]: Could you please add an **aoc** test as well?

Comment (mdtoguchi) [reply clang/test/Driver/sycl-offload.c:688]: Of course - done.

**Pull request merged**